hypothesis-python/src/hypothesis/strategies/_internal/numbers.py

```diff
@@ -183,9 +183,16 @@ def integers(
 
 
 class FloatStrategy(SearchStrategy):
-    """Generic superclass for strategies which produce floats."""
-
-    def __init__(self, allow_infinity, allow_nan, allow_subnormal, width):
+    """A strategy for floating point numbers."""
+
+    def __init__(
+        self,
+        allow_infinity: bool,
+        allow_nan: bool,
+        allow_subnormal: bool,
+        width: int,
+        bounds: Optional[Tuple[float, float]] = None,
```

Comment:
Let's make this a min_value and max_value, defaulting to `0.0` and `+inf` respectively.  We can then be much less conditional everywhere else.

Reply:
I also was unsatisfied with the bimodal-ness of this logic.

So, for one, I think we still need the unbounded FloatStrategy to still be able to return negatives. Also, the way we do bounded selection is also based on [fractional_float](https://github.com/HypothesisWorks/hypothesis/blob/b18a75d477cfb58f7611cb8ff63d1d960f495b49/hypothesis-python/src/hypothesis/strategies/_internal/numbers.py#L269), which I don't think will work with `+inf`. Or it's possible (likely?) that I'm missing the context for why these aren't issues.

One thing I did try - including the appropriate NASTY_FLOATs when bounds are specified. This sort of works, but requires special handling to avoid `-0.0` and a possibly empty list. Then some shrinking tests started failing in a way that made me worry I was causing problems there. Insights appreciated!

Reply:
The underlying FloatStrategy is only positive IIRC, so `st.floats()` has to return a union of two Strategy objects if the allowed range spans zero.

Reply:
It **appears** that the [unbounded case](https://github.com/HypothesisWorks/hypothesis/blob/b18a75d477cfb58f7611cb8ff63d1d960f495b49/hypothesis-python/src/hypothesis/strategies/_internal/numbers.py#L495) doesn't do this union. But this is a big function, and I know that I am not following all of it.

Reply:
Oh, right, yeah, unbounded floats is positive-only *in `conjecture.utils`* (or `conjecture.floats`?  anyway...), but adds a sign bit before you get to `FloatStrategy`.

Reply:
Yeah. Still happy to refactor as you see fit - LMK!

Reply:
Let's push the management of sign-bits-for-bounded-floats down, from `st.floats()` into `FloatsStrategy`.  The high-level goal here is that `FloatsStrategy` should hold enough of the arguments to fully represent a call to `floats()` - so that the solver/fuzzer/filter-rewriting interface will pretty much "wrap the current `FloatsStrategy.do_draw()`".

Kinda annoying now, sorry, but I think it'll pay off in the medium-term.

Reply:
Not annoying!

Ok, so you might imagine roughly [this logic](https://github.com/HypothesisWorks/hypothesis/blob/b18a75d477cfb58f7611cb8ff63d1d960f495b49/hypothesis-python/src/hypothesis/strategies/_internal/numbers.py#L490-L553) getting incorporated into the `FloatStrategy` class. (though needs be implemented rather differently since strategy-composition isn't a tactic that would be as readily available) Is that about right?

Backing up, some questions that come to mind:
1. Would the ideal mid-level interface be identical to `st.floats()`? (FWIW, I think CrossHair could implement st.floats(), FloatStrategy, or anything in between without much trouble) One could argue that some amount of normalization and consistency checking is better off at the higher level.
2. does it matter whether that interface is a function or a class? I'm not familiar with the filter-rewriting stuff, but I imagine that gets a lot easier if this interface is a class-level thing, right? And that's why we care?

Reply:
> Ok, so you might imagine roughly [this logic](https://github.com/HypothesisWorks/hypothesis/blob/b18a75d477cfb58f7611cb8ff63d1d960f495b49/hypothesis-python/src/hypothesis/strategies/_internal/numbers.py#L490-L553) getting incorporated into the `FloatStrategy` class. (though needs be implemented rather differently since strategy-composition isn't a tactic that would be as readily available) Is that about right?

Yep, that sounds right to me.  Caveats: we should check that this doesn't noticably hurt performance in the interim period

> 1. Would the ideal mid-level interface be identical to `st.floats()`? (FWIW, I think CrossHair could implement st.floats(), FloatStrategy, or anything in between without much trouble) One could argue that some amount of normalization and consistency checking is better off at the higher level.

I think **a** mid-level interface I would want is "like `floats()`, but inclusive bounds are required and there's no sign bit or allow_subnormal".  So we're really just talking about finite floats in some contiguous range, plus maybe `nan` (only if `inf` is allowed).  It's tractable to efficiently generate this with bit-twiddling to get close enough to right that at least half of attempts are valid, and that's good enough that rejection sampling is really fast.

It's deeply weird to humans, but this should work really well for fuzzers and pretty well for solvers too.  Then `FloatsStrategy` is responsible for composing positive and negative ranges (if applicable), zero (if  subnormals disallowed), etc.

> 2. does it matter whether that interface is a function or a class? I'm not familiar with the filter-rewriting stuff, but I imagine that gets a lot easier if this interface is a class-level thing, right? And that's why we care?

The filter rewriting will be invoked via `FloatsStrategy.filter()`, so I'd build `FloatsStrategy.do_draw()` directly in terms of the mid-level interface.  I'd imagined that the interface is basically "provide an object with these callbacks", which, yeah, is very likely to be an instance of some class with those methods.